File: 2374-steps-to-make-array-non-decreasing/steps-to-make-array-non-decreasing.java
import java.util.*;

class Solution {
    public int totalSteps(int[] nums) {

        Stack<int[]> stack = new Stack<>();
        int ans = 0;

        for (int num : nums) {

            int steps = 0;

            while (!stack.isEmpty() && num >= stack.peek()[0]) {
                steps = Math.max(steps, stack.peek()[1]);
                stack.pop();
            }

            if (!stack.isEmpty()) {
                steps++;
            } else {
                steps = 0;
            }

            ans = Math.max(ans, steps);

            stack.push(new int[]{num, steps});
        }

        return ans;
    }
}

File: 2374-steps-to-make-array-non-decreasing/steps-to-make-array-non-decreasing.py
class Solution:
    def totalSteps(self, nums):
        stack = []
        ans = 0

        for num in nums:
            steps = 0

            while stack and num >= stack[-1][0]:
                steps = max(steps, stack[-1][1])
                stack.pop()

            if stack:
                steps += 1
            else:
                steps = 0

            ans = max(ans, steps)
            stack.append((num, steps))

        return ans
        